### PR TITLE
[SPARK-31290][R] Add back the deprecated R APIs

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add back the deprecated R APIs removed by https://github.com/apache/spark/pull/22843/ and https://github.com/apache/spark/pull/22815.

These APIs are

- `sparkR.init`
- `sparkRSQL.init`
- `sparkRHive.init`
- `registerTempTable`
- `createExternalTable`
- `dropTempTable`

No need to port the function such as 
```r
createExternalTable <- function(x, ...) {
  dispatchFunc("createExternalTable(tableName, path = NULL, source = NULL, ...)", x, ...)
}
```
because this was for the backward compatibility when SQLContext exists before assuming from https://github.com/apache/spark/pull/9192,  but seems we don't need it anymore since SparkR replaced SQLContext with Spark Session at https://github.com/apache/spark/pull/13635.

### Why are the changes needed?
Amend Spark's Semantic Versioning Policy

### Does this PR introduce any user-facing change?
Yes
The removed R APIs are put back.

### How was this patch tested?
Add back the removed tests

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @HyukjinKwon @felixcheung

**Comment by huaxingao** [comment]
Do we need to add back ```saveAsParquetFile```? Seems it is not added back in scala? Also, do we need to add back this ```sparkR.init```, ```sparkRSQL.init```, ```sparkRHive.init```?

**Comment by gatorsmile** [comment]
Also cc @mengxr @falaki Could you take a look?

**Comment by falaki** [comment]
If we have added these back to Scala, adding them to R seems good.

**Comment by dongjoon-hyun** [comment]
cc @marmbrus

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -946,6 +973,17 @@ setMethod("write.parquet",
             invisible(handledCallJMethod(write, "parquet", path))
           })
 
+#' @rdname write.parquet
+#' @name saveAsParquetFile
+#' @aliases saveAsParquetFile,SparkDataFrame,character-method
+#' @note saveAsParquetFile since 1.4.0
+setMethod("saveAsParquetFile",
```
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R]: Are we sure we want to keep this too? Scala and PySpark sides don't have it either as of SPARK-12600 cc @rxin and @marmbrus 

https://github.com/apache/spark/commit/77ab49b8575d2ebd678065fa70b0343d532ab9c2#diff-c3d0394b2fc08fb2842ff0362a5ac6c9

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R]: `jsonFile` and `parquetFile` too.

**Review by HyukjinKwon — APPROVED** [review]
LGTM except one question https://github.com/apache/spark/pull/28058/files#r399743329

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R]: Okay, using the rubric we added, in R or PySpark, it's very easy to keep the compatibility - if-else based on version and redirect to the alternative. There seems direct alternative too.

Seems like we haven't seen a lot of complaints when we remove `saveAsParquetFile` at  SPARK-12600. I skimmed user-mailing list with keyword search, and I found 0 complaints.

Having write API in `DataFrame` seems inconsistent with Scala and PySpark sides, and also namespace used in R side (e.g., `write.json`) - it causes a cognitive cost.

Shall we remove these one out?

**Comment by felixcheung** [comment]
They don’t have 1:1 to Scala. Some R APIs are older and long since removed even in Scala before 3.0.

I’m fine with this change if this is what the community has agreed on.

**Comment by rxin** [comment]
If they have already been removed prior to 3.0 and nobody has said anything, I don't think we should add those back in.

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (gatorsmile) [reply R/pkg/R/DataFrame.R]: OK. If pySpark already removed it in 2.0, I think we can get rid of saveAsParquetFile in sparkR in this release too.

**Comment by HyukjinKwon** [comment]
Just to clarify things, what the community agreed on is the rubric added at https://spark.apache.org/versioning-policy.html.

Some APIs listed here, `jsonFile`, `parquetFile` and `saveAsParquetFile` were removed in Scala and PySpark before Spark 3.0. I skimmed nobody complained in user mailing list by keyword search.

One last thing is, SparkR dropped RDD support. So I don't think it also makes sense to add `jsonRDD`.

**Comment by HyukjinKwon** [comment]
@huaxingao, seems like we roughly agreed upon excluding `jsonFile`, `parquetFile` and `saveAsParquetFile`. `jsonRDD` seems not making sense to add back since SparkR dropped RDD support officially. Can we remove these four in this PR?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@huaxingao, don't forget to update PR description :-)

**Review by HyukjinKwon — COMMENTED** [review]
@huaxingao, can you also add the reasons why we didn't port:

```r
createExternalTable <- function(x, ...) {
  dispatchFunc("createExternalTable(tableName, path = NULL, source = NULL, ...)", x, ...)
}
```

in the PR description?

Apparently, this was for the backward compatibility when SQLContext exists before assuming from https://github.com/apache/spark/pull/9192 but seems we don't need it anymore since SparkR replaced SQLContext with Spark Session at https://github.com/apache/spark/pull/13635

File: R/pkg/R/catalog.R [hunk 2]
```diff
@@ -130,6 +159,29 @@ clearCache <- function() {
   invisible(callJMethod(catalog, "clearCache"))
 }
 
+#' Drops the temporary table with the given table name in the catalog.
```
Comment (HyukjinKwon) [inline R/pkg/R/catalog.R]: `#' (Deprecated) Drop Temporary Table` seems missed.

**Comment by HyukjinKwon** [comment]
It seems more complicated than I though. Let me try to clarify the current status here:

#### `jsonFile` and `parquetFile`

- Scala side, `jsonFile`, `parquetFile` were removed at https://github.com/apache/spark/pull/10559 and recovered at https://github.com/apache/spark/pull/13637 in the same Spark 2.0.
- PySpark side, they were removed at https://github.com/apache/spark/pull/10559 and were not recovered afterwards.
- SparkR side, they were never removed

#### `jsonRDD`

- SparkR made RDD API as private APIs as of SPARK-7230; RDD related APIs are supposed to be removed as they are uesless.

#### `saveAsParquetFile`

- Scala side `saveAsParquetFile` was removed at https://github.com/apache/spark/pull/10559 
- PySpark side, they were removed at https://github.com/apache/spark/pull/10559.
- SparkR side, they were never removed.

I skimmed user-mailing list with keyword search, and I found no complaints. This implies, no complaints found about:
 -  `jsonFile`, `parquetFile` and `jsonRDD` missing in PySpark from Spark 2.0
 -  `saveAsParquetFile` in Scala and PySpark from Spark 2.0

So, I guess it's fine to don't add `jsonFile`, `parquetFile`, `jsonRDD` and `saveAsParquetFile` back.

<br/>

TR;DL: from my rough investigation, I think it's fine to don't add `jsonFile`, `parquetFile`, `jsonRDD` and `saveAsParquetFile` because:
  - people don't quite care about `jsonFile`, `parquetFile`, `jsonRDD` and `saveAsParquetFile`. 
  - `jsonFile` and `parquetFile` don't exist in PySpark.
  - `jsonRDD` is useless as RDD APIs became private in SparkR.
  - `saveAsParquetFile` does not exist in Scala and PySpark sides.

WDYT @falaki and @mengxr?

**Comment by HyukjinKwon** [comment]
I discussed offline with @marmbrus. Looks we're fine to exclude these four above.

Merged to master and branch-3.0.

**Comment by huaxingao** [comment]
Thank you all for the help!

**Pull request closed without merging**